File: packages/java/src/main/java/com/logolabs/inkvec/Inkvec.java
package com.logolabs.inkvec;

import com.sun.jna.Native;
import com.sun.jna.Pointer;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Objects;

/**
 * Inkvec: raster logos, icons and illustrations to compact, accurate SVG.
 *
 * <p>This is the Java binding over Inkvec's C ABI ({@code include/inkvec.h}), loaded through
 * <a href="https://github.com/java-native-access/jna">JNA</a>. It contains no per-option
 * logic: options are either a generated, typed {@link InkvecOptions} or a raw JSON string,
 * and the native library ({@code inkvec::Options::from_json}) is the only place a value is
 * defaulted or validated. Every method may be called from any number of threads at once.
 *
 * <pre>{@code
 * byte[] png = Files.readAllBytes(Path.of("logo.png"));
 * TraceResult r = Inkvec.trace(png, InkvecOptions.builder().colors(16).build());
 * Files.writeString(Path.of("logo.svg"), r.svg());
 * }</pre>
 *
 * <h2>Loading the native library</h2>
 *
 * The library name is {@code inkvec_ffi}. JNA looks for it, in order: the system property
 * {@code jna.library.path} (a shortcut, the system property {@code inkvec.library.path}, is
 * folded into it if set before this class loads), the classpath under JNA's per-platform
 * resource prefix ({@code win32-x86-64/inkvec_ffi.dll}, {@code linux-x86-64/libinkvec_ffi.so},
 * {@code linux-aarch64/...}, {@code darwin-x86-64/...}, {@code darwin-aarch64/...} -- where a
 * release jar bundles it), then the platform's normal library search path.
 */
public final class Inkvec {

    private static final int INKVEC_OK = 0;
    private static final int INKVEC_ERR_INVALID_ARGUMENT = 1;
    private static final int INKVEC_ERR_INVALID_IMAGE = 2;
    private static final int INKVEC_ERR_INVALID_OPTIONS = 3;
    private static final int INKVEC_ERR_INTERNAL = 4;

    private static final InkvecLibrary NATIVE;

    static {
        // A project-specific alias for jna.library.path, folded in before JNA resolves the
        // library, so a caller need not know JNA's own property name.
        String extra = System.getProperty("inkvec.library.path");
        if (extra != null && !extra.isEmpty()) {
            String existing = System.getProperty("jna.library.path", "");
            String merged = existing.isEmpty() ? extra : existing + java.io.File.pathSeparator + extra;
            System.setProperty("jna.library.path", merged);
        }
        // Marshal every String (options JSON) as UTF-8, independent of the platform default
        // charset; SVG and error text go through explicit Pointer decoding below regardless.
        if (System.getProperty("jna.encoding") == null) {
            System.setProperty("jna.encoding", "UTF-8");
        }
        NATIVE = Native.load("inkvec_ffi", InkvecLibrary.class);
    }

    private Inkvec() {
    }

    /** Trace an encoded image (PNG, JPEG, WebP, GIF, BMP or TIFF) with every option at its default. */
    public static TraceResult trace(byte[] image) {
        return trace(image, (String) null);
    }

    /** Trace an encoded image with the given typed options. */
    public static TraceResult trace(byte[] image, InkvecOptions options) {
        return trace(image, options.toJson());
    }

    /**
     * Trace an encoded image with options as a raw JSON object -- an escape hatch for an
     * option added to Inkvec before {@link InkvecOptions} is regenerated for it.
     */
    public static TraceResult trace(byte[] image, String optionsJson) {
        Objects.requireNonNull(image, "image");
        InkvecResultStruct out = new InkvecResultStruct();
        try {
            int status = NATIVE.inkvec_trace(image, image.length, optionsJson, out);
            return result(status, out);
        } finally {
            NATIVE.inkvec_result_free(out);
        }
    }

    /** Reads the whole file, then traces it with every option at its default. */
    public static TraceResult trace(Path path) throws IOException {
        return trace(Files.readAllBytes(path), (String) null);
    }

    /** Reads the whole file, then traces it with the given typed options. */
    public static TraceResult trace(Path path, InkvecOptions options) throws IOException {
        return trace(Files.readAllBytes(path), options.toJson());
    }

    /** Reads the whole file, then traces it with options as a raw JSON object. */
    public static TraceResult trace(Path path, String optionsJson) throws IOException {
        return trace(Files.readAllBytes(path), optionsJson);
    }

    /**
     * Trace raw pixels: straight (not premultiplied) RGBA, 8 bits per channel, row-major,
     * tightly packed, exactly {@code width * height * 4} bytes. Every option at its default.
     * Byte-identical to {@link #trace(byte[])} on a PNG holding the same pixels.
     */
    public static TraceResult traceRgba(byte[] rgba, int width, int height) {
        return traceRgba(rgba, width, height, (String) null);
    }

    /** Trace raw RGBA pixels with the given typed options. */
    public static TraceResult traceRgba(byte[] rgba, int width, int height, InkvecOptions options) {
        return traceRgba(rgba, width, height, options.toJson());
    }

    /** Trace raw RGBA pixels with options as a raw JSON object. */
    public static TraceResult traceRgba(byte[] rgba, int width, int height, String optionsJson) {
        Objects.requireNonNull(rgba, "rgba");
        InkvecResultStruct out = new InkvecResultStruct();
        try {
            int status = NATIVE.inkvec_trace_rgba(rgba, rgba.length, width, height, optionsJson, out);
            return result(status, out);
        } finally {
            NATIVE.inkvec_result_free(out);
        }
    }

    /** The JSON Schema (draft 2020-12) of the options: every name, type, default, range. */
    public static String optionsSchema() {
        return decode(NATIVE.inkvec_options_schema());
    }

    /** Every option at its default, as a compact JSON object. */
    public static String defaults() {
        return decode(NATIVE.inkvec_default_options());
    }

    /** The library version, e.g. {@code "0.1.3"} -- read from the loaded native library. */
    public static String version() {
        return decode(NATIVE.inkvec_version());
    }

    /**
     * The target the native library was compiled for, e.g. {@code "x86_64-windows-msvc"}.
     * Output is byte-identical between builds with the same target; another target can write
     * the same drawing slightly differently (see docs/BINDINGS.md, "Determinism").
     */
    public static String buildTarget() {
        return decode(NATIVE.inkvec_build_target());
    }

    /** The C ABI version of the loaded native library, for a compatibility check at startup. */
    public static int abiVersion() {
        return NATIVE.inkvec_abi_version();
    }

    private static TraceResult result(int status, InkvecResultStruct out) {
        if (status == INKVEC_OK) {
            // svg_len is authoritative (the header calls svg NUL-terminated too, but this
            // does not depend on that): read exactly that many bytes and decode as UTF-8.
            byte[] bytes = out.svg.getByteArray(0, (int) out.svg_len);
            String svg = new String(bytes, StandardCharsets.UTF_8);
            return new TraceResult(svg, out.width, out.height);
        }
        String message = out.error == null ? "" : out.error.getString(0, StandardCharsets.UTF_8.name());
        switch (status) {
            case INKVEC_ERR_INVALID_IMAGE:
                throw new InvalidImageException(message);
            case INKVEC_ERR_INVALID_OPTIONS:
                throw new InvalidOptionsException(message);
            case INKVEC_ERR_INTERNAL:
                throw new InternalException(message);
            case INKVEC_ERR_INVALID_ARGUMENT:
                throw new IllegalArgumentException(message);
            default:
                throw new InternalException("unknown status " + status + ": " + message);
        }
    }

    private static String decode(Pointer p) {
        return p.getString(0, StandardCharsets.UTF_8.name());
    }
}

File: packages/php/src/Inkvec.php
<?php

declare(strict_types=1);

namespace LogoLabs\Inkvec;

use FFI;

/**
 * Inkvec: raster logos, icons and illustrations to compact, accurate SVG.
 *
 * This is the PHP binding over Inkvec's C ABI (`crates/inkvec-ffi`), called through ext-FFI
 * -- the tracer itself runs as native code in this process, with no subprocess, no HTTP hop
 * and no copy of the image on the way in. It holds no per-option logic: options travel as
 * one JSON object, and the native library (`inkvec::Options::from_json`) is the only place a
 * value is defaulted or validated.
 *
 *     $traced = Inkvec::traceFile('logo.png', new Options(colors: 16));
 *     file_put_contents('logo.svg', $traced->svg);
 *
 * Every method may be called as often as you like; the library is loaded once per process
 * (see {@see preload()} for pools) and is thread-safe, though PHP itself is not threaded.
 */
final class Inkvec
{
    private function __construct()
    {
    }

    /**
     * Trace an encoded image: the bytes of a PNG, JPEG, WebP, GIF, BMP or TIFF.
     *
     * @param string                    $image   the file's bytes, not a path
     * @param Options|array|string|null $options typed options, an array keyed by the
     *                                           tracer's own option names, a raw JSON
     *                                           object, or null for every default
     *
     * @throws InvalidImageException   the bytes are not an image Inkvec can decode
     * @throws InvalidOptionsException an unknown option, or a value of the wrong type or out of range
     * @throws InternalException       the tracer failed; worth a bug report
     * @throws LibraryException        the native library could not be loaded
     */
    public static function trace(string $image, Options|array|string|null $options = null): Traced
    {
        $ffi = Library::ffi();
        $out = $ffi->new('InkvecResult');
        $out->struct_size = FFI::sizeof($out);

        try {
            $status = $ffi->inkvec_trace($image, \strlen($image), self::optionsJson($options), FFI::addr($out));

            return self::result($status, $out);
        } finally {
            $ffi->inkvec_result_free(FFI::addr($out));
        }
    }

    /**
     * Read a file and trace it. The whole file is read into memory first, as the tracer
     * works on bytes.
     *
     * @param Options|array|string|null $options see {@see trace()}
     *
     * @throws InkvecException if the file cannot be read, plus everything {@see trace()} throws
     */
    public static function traceFile(string $path, Options|array|string|null $options = null): Traced
    {
        $bytes = @file_get_contents($path);
        if ($bytes === false) {
            throw new InkvecException("inkvec: cannot read {$path}");
        }

        return self::trace($bytes, $options);
    }

    /**
     * Trace raw pixels: straight (not premultiplied) RGBA, 8 bits per channel, row-major,
     * tightly packed, exactly `width * height * 4` bytes. The SVG is byte-identical to
     * {@see trace()} on a PNG holding the same pixels.
     *
     * Pixels that do not match the size given are an {@see InvalidImageException}, raised by
     * the tracer, not counted here.
     *
     * @param Options|array|string|null $options see {@see trace()}
     */
    public static function traceRgba(
        string $rgba,
        int $width,
        int $height,
        Options|array|string|null $options = null
    ): Traced {
        $ffi = Library::ffi();
        $out = $ffi->new('InkvecResult');
        $out->struct_size = FFI::sizeof($out);

        try {
            $status = $ffi->inkvec_trace_rgba(
                $rgba,
                \strlen($rgba),
                $width,
                $height,
                self::optionsJson($options),
                FFI::addr($out)
            );

            return self::result($status, $out);
        } finally {
            $ffi->inkvec_result_free(FFI::addr($out));
        }
    }

    /** The native library's version, e.g. `0.1.4`. */
    public static function version(): string
    {
        return Library::ffi()->inkvec_version();
    }

    /**
     * The target the native library was built for, e.g. `x86_64-linux-gnu`. Output is
     * byte-identical between builds for one target; another target can write the same
     * drawing slightly differently (docs/BINDINGS.md, "Determinism").
     */
    public static function buildTarget(): string
    {
        return Library::ffi()->inkvec_build_target();
    }

    /** The C ABI version of the loaded library; always {@see Library::ABI_VERSION} here. */
    public static function abiVersion(): int
    {
        return Library::ffi()->inkvec_abi_version();
    }

    /**
     * The JSON Schema (draft 2020-12) of the options: every name, type, default, description
     * and range, read from the library itself, so it can never drift from it.
     *
     * @return array<string, mixed>
     */
    public static function optionsSchema(): array
    {
        return json_decode(self::optionsSchemaJson(), true, 512, \JSON_THROW_ON_ERROR);
    }

    /** The same schema, as the library's own JSON text. */
    public static function optionsSchemaJson(): string
    {
        return Library::ffi()->inkvec_options_schema();
    }

    /**
     * Every option at its default.
     *
     * @return array<string, bool|int|float>
     */
    public static function defaults(): array
    {
        return json_decode(self::defaultsJson(), true, 512, \JSON_THROW_ON_ERROR);
    }

    /** The same defaults, as the library's own compact JSON object. */
    public static function defaultsJson(): string
    {
        return Library::ffi()->inkvec_default_options();
    }

    /**
     * Whether a trace would work here: ext-FFI present and a native library that loads.
     * Useful to pick between tracing in-process and some other route at runtime.
     */
    public static function available(): bool
    {
        try {
            Library::ffi();

            return true;
        } catch (LibraryException) {
            return false;
        }
    }

    /**
     * Load this native library instead of searching for one. Pass null to forget it. Must
     * be called before the first trace of the process.
     *
     * @see Library::candidates() for what is searched otherwise
     */
    public static function useLibrary(?string $path): void
    {
        Library::use($path);
    }

    /**
     * The library in use, or null when none has been loaded yet -- and also when it was
     * preloaded, since the path was resolved in the preloading process, not this request
     * ({@see preloaded()}).
     */
    public static function libraryPath(): ?string
    {
        return Library::path();
    }

    /** Whether this process traces through a preloaded scope rather than its own load. */
    public static function preloaded(): bool
    {
        return Library::isPreloaded();
    }

    /**
     * Open the library once for a whole PHP-FPM pool, from an `opcache.preload` script:
     *
     *     require __DIR__ . '/vendor/autoload.php';
     *     LogoLabs\Inkvec\Inkvec::preload();
     *
     * Requests then reach it through an FFI scope, which is also what makes tracing possible
     * when `ffi.enable=preload` (the safe setting for a web SAPI). Returns the path of the
     * generated FFI header.
     */
    public static function preload(?string $library = null, ?string $headerDir = null): string
    {
        return Library::preload($library, $headerDir);
    }

    /**
     * Options as the C ABI takes them: one JSON object, or null for the defaults. An array
     * is encoded as it stands -- its keys are the tracer's own option names (`no_background`,
     * not `noBackground`) -- and a string is passed through untouched, so an option added to
     * Inkvec is reachable before {@see Options} has been regenerated for it. Either way the
     * native library is what validates the result.
     */
    private static function optionsJson(Options|array|string|null $options): ?string
    {
        if ($options === null) {
            return null;
        }
        if ($options instanceof Options) {
            return $options->toJson();
        }
        if (\is_array($options)) {
            return $options === []
                ? '{}'
                : json_encode($options, \JSON_THROW_ON_ERROR | \JSON_PRESERVE_ZERO_FRACTION);
        }

        return trim($options) === '' ? null : $options;
    }

    /**
     * One filled `InkvecResult` as a {@see Traced}, or the matching exception. The strings
     * are copied out of the library's memory here; the caller frees the result either way.
     *
     * @param \FFI\CData $out
     */
    private static function result(int $status, $out): Traced
    {
        if ($status === Library::OK) {
            return new Traced(
                $out->svg_len > 0 ? FFI::string($out->svg, $out->svg_len) : '',
                $out->width,
                $out->height
            );
        }

        $message = $out->error !== null
            ? FFI::string($out->error)
            : "the tracer failed with status {$status}";

        throw match ($status) {
            Library::ERR_INVALID_IMAGE => new InvalidImageException($message),
            Library::ERR_INVALID_OPTIONS => new InvalidOptionsException($message),
            Library::ERR_INTERNAL => new InternalException($message),
            default => new InkvecException($message),
        };
    }
}
